transports/tls/src/verifier.rs

```diff
@@ -0,0 +1,242 @@
+/// - The certificate must have a valid libp2p extension that includes a
+///   signature of its public key.
+impl Libp2pCertificateVerifier {
+    pub fn new() -> Self {
+        Self {
+            remote_peer_id: None,
+        }
+    }
+    pub fn with_remote_peer_id(remote_peer_id: Option<PeerId>) -> Self {
+        Self { remote_peer_id }
+    }
+
+    /// Return the list of SignatureSchemes that this verifier will handle,
+    /// in `verify_tls12_signature` and `verify_tls13_signature` calls.
+    ///
+    /// This should be in priority order, with the most preferred first.
+    pub fn verification_schemes() -> Vec<SignatureScheme> {
+        vec![
+            // TODO SignatureScheme::ECDSA_NISTP521_SHA512 is not supported by `ring` yet
+            SignatureScheme::ECDSA_NISTP384_SHA384,
+            SignatureScheme::ECDSA_NISTP256_SHA256,
+            // TODO SignatureScheme::ED448 is not supported by `ring` yet
+            SignatureScheme::ED25519,
+            // In particular, RSA SHOULD NOT be used unless
+            // no elliptic curve algorithms are supported.
+            SignatureScheme::RSA_PSS_SHA512,
+            SignatureScheme::RSA_PSS_SHA384,
+            SignatureScheme::RSA_PSS_SHA256,
+            SignatureScheme::RSA_PKCS1_SHA512,
+            SignatureScheme::RSA_PKCS1_SHA384,
+            SignatureScheme::RSA_PKCS1_SHA256,
+        ]
+    }
+}
+
+impl ServerCertVerifier for Libp2pCertificateVerifier {
+    fn verify_server_cert(
+        &self,
+        end_entity: &Certificate,
+        intermediates: &[Certificate],
+        _server_name: &rustls::ServerName,
+        _scts: &mut dyn Iterator<Item = &[u8]>,
+        _ocsp_response: &[u8],
+        _now: std::time::SystemTime,
+    ) -> Result<ServerCertVerified, rustls::Error> {
+        let peer_id = verify_presented_certs(end_entity, intermediates)?;
+
+        if let Some(remote_peer_id) = self.remote_peer_id {
+            // The public host key allows the peer to calculate the peer ID of the peer
+            // it is connecting to. Clients MUST verify that the peer ID derived from
+            // the certificate matches the peer ID they intended to connect to,
+            // and MUST abort the connection if there is a mismatch.
+            if remote_peer_id != peer_id {
+                return Err(rustls::Error::PeerMisbehavedError(
+                    "Wrong peer ID in p2p extension".to_string(),
+                ));
+            }
+        }
+
+        Ok(ServerCertVerified::assertion())
+    }
+
+    fn verify_tls12_signature(
+        &self,
+        _message: &[u8],
+        _cert: &Certificate,
+        _dss: &DigitallySignedStruct,
+    ) -> Result<HandshakeSignatureValid, rustls::Error> {
+        // The libp2p handshake uses TLS 1.3 (and higher).
+        // Endpoints MUST NOT negotiate lower TLS versions.
+        Err(rustls::Error::PeerIncompatibleError(
+            "Only TLS 1.3 certificates are supported".to_string(),
```

Comment:
Why is this check done when verifying the certificate? The TLS version is negotiated before certificates are exchanged.

I also believe that the term "TLS 1.3 certificate" is misleading. Certificates are not bound to the TLS version.

Reply:
> The TLS version is negotiated before certificates are exchanged.

That's right, so this is a no-op just in case.

Reply:
> Why is this check done when verifying the certificate? The TLS version is negotiated before certificates are exchanged.

The TLS library at hand supports multiple TLS versions but they are all combined into one interface. Meaning at compile-time, we don't know which TLS version will be used and thus we need to satisfy this interface and provide an implementation for `verify_tls12_signature`. This function will should never be called because we disable anything other than 1.3 at configuration time. Due to TLS' flexibility, this is hard to express statically so I assume the authors of the TLS library didn't bother in modelling this using the type system.

> I also believe that the term "TLS 1.3 certificate" is misleading. Certificates are not bound to the TLS version.

I'll change the error message.